libraries/AP_Math/AP_Math.cpp

```diff
@@ -124,60 +129,57 @@ float throttle_curve(float thr_mid, float alpha, float thr_in)
 }
 
 template <typename T>
-float wrap_180(const T angle, float unit_mod)
+T wrap_180(const T angle, T unit_mod)
 {
     auto res = wrap_360(angle, unit_mod);
-    if (res > 180.f * unit_mod) {
-        res -= 360.f * unit_mod;
+    if (res > T(180) * unit_mod) {
+        res -= T(360) * unit_mod;
     }
     return res;
 }
 
-template float wrap_180<int>(const int angle, float unit_mod);
-template float wrap_180<short>(const short angle, float unit_mod);
+template int wrap_180<int>(const int angle, int unit_mod);
+template short wrap_180<short>(const short angle, short unit_mod);
 template float wrap_180<float>(const float angle, float unit_mod);
-template float wrap_180<double>(const double angle, float unit_mod);
+template double wrap_180<double>(const double angle, double unit_mod);
 
 template <typename T>
-auto wrap_180_cd(const T angle) -> decltype(wrap_180(angle, 100.f))
+auto wrap_180_cd(const T angle) -> decltype(wrap_180(angle, T(100)))
 {
-    return wrap_180(angle, 100.f);
+    return wrap_180(angle, T(100));
 }
 
-template auto wrap_180_cd<float>(const float angle) -> decltype(wrap_180(angle, 100.f));
-template auto wrap_180_cd<int>(const int angle) -> decltype(wrap_180(angle, 100.f));
-template auto wrap_180_cd<long>(const long angle) -> decltype(wrap_180(angle, 100.f));
-template auto wrap_180_cd<short>(const short angle) -> decltype(wrap_180(angle, 100.f));
-template auto wrap_180_cd<double>(const double angle) -> decltype(wrap_360(angle, 100.f));
+template auto wrap_180_cd<float>(const float angle) -> decltype(wrap_180(angle, float(100.f)));
+template auto wrap_180_cd<int>(const int angle) -> decltype(wrap_180(angle, int(100)));
+template auto wrap_180_cd<long>(const long angle) -> decltype(wrap_180(angle, long(100)));
+template auto wrap_180_cd<double>(const double angle) -> decltype(wrap_360(angle, double(100)));
 
 template <typename T>
-float wrap_360(const T angle, float unit_mod)
+T wrap_360(const T angle, T unit_mod)
 {
-    const float ang_360 = 360.f * unit_mod;
-    float res = fmodf(static_cast<float>(angle), ang_360);
+    const auto ang_360 = T(360) * unit_mod;
+    auto res = fmodf(static_cast<T>(angle), ang_360);
```

Comment:
The result of this is still going to be float, and the arguments are all going to be cast back to float anyways before calling it, so I don't think anything has actually been improved yet. (Arguably it's worse because now we both cast into and out of float, where as before we could sometimes stay in the float realm)